File: front/src/api/group.ts
import http from './http'
import type { ApiResponse, PageResponse } from './chat'

export interface GroupSummary {
  groupId: number
  ownerId: number
  groupName: string
  description: string | null
  avatarUrl: string | null
  notice: string | null
  status: number
  inviteCode: string | null
  chatEnabled: number
  recallLimitMinutes: number
  memberCount: number
  unreadCount: number
  mentionUnreadCount: number
  noticeUnread: boolean
  noticeUpdatedAt: string | null
  createdAt: string
  updatedAt: string
}

export interface GroupMember {
  userId: number
  email: string
  nickname: string
  avatarUrl: string | null
  role: number
  groupNickname: string | null
  muteUntil: string | null
  joinedAt: string
}

export interface GroupMessage {
  messageId: number
  groupId: number
  senderId: number
  senderNickname: string
  senderAvatarUrl: string | null
  messageType: 'text' | 'image' | 'file'
  content: string
  fileId: number | null
  replyToMessageId: number | null
  replyPreview: string | null
  mentionUserIds: number[]
  fileName: string | null
  fileMimeType: string | null
  fileSize: number | null
  mentionAll: number
  recallStatus: number
  sentAt: string
  updatedAt: string
}

export interface GroupJoinRequest {
  requestId: number
  groupId: number
  requesterId: number
  requesterEmail: string
  requesterNickname: string
  requesterAvatarUrl: string | null
  message: string | null
  status: number
  createdAt: string
}

export interface GroupNotification {
  notificationId: number
  groupId: number
  actorId: number | null
  targetUserId: number | null
  noticeType: string
  content: string
  createdAt: string
}

export interface GroupNoticeReadStats {
  readCount: number
  memberCount: number
}

export async function createGroup(groupName: string, description: string, memberIds: number[], avatarUrl?: string) {
  const response = await http.post<ApiResponse<GroupSummary>>('/groups', {
    groupName,
    description,
    memberIds,
    avatarUrl,
  })
  return response.data.data
}

export async function listGroups() {
  const response = await http.get<ApiResponse<GroupSummary[]>>('/groups')
  return response.data.data
}

export async function getGroup(groupId: number) {
  const response = await http.get<ApiResponse<GroupSummary>>(`/groups/${groupId}`)
  return response.data.data
}

export async function searchGroups(keyword: string, page = 1, pageSize = 20) {
  const response = await http.get<ApiResponse<PageResponse<GroupSummary>>>('/groups/search', {
    params: { keyword, page, pageSize },
  })
  return response.data.data
}

export async function updateGroupProfile(
  groupId: number,
  payload: {
    groupName?: string
    description?: string
    avatarUrl?: string
    chatEnabled?: boolean
    recallLimitMinutes?: number
  },
) {
  const response = await http.put<ApiResponse<GroupSummary>>(`/groups/${groupId}`, payload)
  return response.data.data
}

export async function joinGroupByInviteCode(inviteCode: string) {
  const response = await http.post<ApiResponse<GroupSummary>>(`/groups/invite/${inviteCode}/join`)
  return response.data.data
}

export async function requestJoinGroup(groupId: number, message = '') {
  const response = await http.post<ApiResponse<GroupJoinRequest>>(`/groups/${groupId}/join-requests`, { message })
  return response.data.data
}

export async function listGroupJoinRequests(groupId: number) {
  const response = await http.get<ApiResponse<GroupJoinRequest[]>>(`/groups/${groupId}/join-requests`)
  return response.data.data
}

export async function listMyGroupJoinRequests() {
  const response = await http.get<ApiResponse<GroupJoinRequest[]>>('/groups/join-requests/me')
  return response.data.data
}

export async function approveGroupJoinRequest(groupId: number, requestId: number) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/join-requests/${requestId}/approve`)
  return response.data.data
}

export async function rejectGroupJoinRequest(groupId: number, requestId: number) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/join-requests/${requestId}/reject`)
  return response.data.data
}

export async function listGroupNotifications(groupId: number, page = 1, pageSize = 20) {
  const response = await http.get<ApiResponse<PageResponse<GroupNotification>>>(`/groups/${groupId}/notifications`, {
    params: { page, pageSize },
  })
  return response.data.data
}

export async function leaveGroup(groupId: number) {
  await http.post(`/groups/${groupId}/leave`)
}

export async function dissolveGroup(groupId: number) {
  await http.delete(`/groups/${groupId}`)
}

export async function transferGroupOwner(groupId: number, targetUserId: number) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/owner`, { targetUserId })
  return response.data.data
}

export async function listGroupMembers(groupId: number) {
  const response = await http.get<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members`)
  return response.data.data
}

export async function listGroupMessages(groupId: number, page = 1, pageSize = 30) {
  const response = await http.get<ApiResponse<PageResponse<GroupMessage>>>(`/groups/${groupId}/messages`, {
    params: { page, pageSize },
  })
  return response.data.data
}

export async function listGroupFiles(groupId: number, page = 1, pageSize = 20) {
  const response = await http.get<ApiResponse<PageResponse<GroupMessage>>>(`/groups/${groupId}/files`, {
    params: { page, pageSize },
  })
  return response.data.data
}

export async function markGroupAsRead(groupId: number) {
  await http.post(`/groups/${groupId}/read`)
}

export async function searchGroupMessages(groupId: number, keyword: string, page = 1, pageSize = 20) {
  const response = await http.get<ApiResponse<PageResponse<GroupMessage>>>(`/groups/${groupId}/messages/search`, {
    params: { keyword, page, pageSize },
  })
  return response.data.data
}

export async function sendGroupMessage(
  groupId: number,
  content: string,
  messageType: GroupMessage['messageType'] = 'text',
  fileId?: number,
  mentionAll = false,
  replyToMessageId?: number,
  mentionUserIds: number[] = [],
) {
  const response = await http.post<ApiResponse<GroupMessage>>(`/groups/${groupId}/messages`, {
    messageType,
    content,
    fileId,
    mentionAll,
    replyToMessageId,
    mentionUserIds,
  })
  return response.data.data
}

export async function sendGroupMentionAllMessage(groupId: number, content: string, replyToMessageId?: number) {
  const response = await http.post<ApiResponse<GroupMessage>>(`/groups/${groupId}/messages/mention-all`, {
    messageType: 'text',
    content,
    replyToMessageId,
  })
  return response.data.data
}

export async function recallGroupMessage(groupId: number, messageId: number) {
  await http.post(`/groups/${groupId}/messages/${messageId}/recall`)
}

export async function deleteGroupMessage(groupId: number, messageId: number) {
  await http.delete(`/groups/${groupId}/messages/${messageId}`)
}

export async function addGroupMembers(groupId: number, memberIds: number[]) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members`, { memberIds })
  return response.data.data
}

export async function updateMyGroupNickname(groupId: number, nickname: string) {
  const response = await http.put<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members/me/nickname`, { nickname })
  return response.data.data
}

export async function removeGroupMember(groupId: number, userId: number) {
  await http.delete(`/groups/${groupId}/members/${userId}`)
}

export async function updateGroupAnnouncement(groupId: number, notice: string) {
  const response = await http.post<ApiResponse<GroupSummary>>(`/groups/${groupId}/announcement`, { notice })
  return response.data.data
}

export async function markGroupAnnouncementRead(groupId: number) {
  const response = await http.post<ApiResponse<GroupNoticeReadStats>>(`/groups/${groupId}/announcement/read`)
  return response.data.data
}

export async function getGroupAnnouncementReadStats(groupId: number) {
  const response = await http.get<ApiResponse<GroupNoticeReadStats>>(`/groups/${groupId}/announcement/read-stats`)
  return response.data.data
}

export async function setGroupAdmin(groupId: number, userId: number) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members/${userId}/admin`)
  return response.data.data
}

export async function unsetGroupAdmin(groupId: number, userId: number) {
  const response = await http.delete<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members/${userId}/admin`)
  return response.data.data
}

export async function muteGroupMember(groupId: number, userId: number, minutes: number) {
  const response = await http.post<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members/${userId}/mute`, { minutes })
  return response.data.data
}

export async function unmuteGroupMember(groupId: number, userId: number) {
  const response = await http.delete<ApiResponse<GroupMember[]>>(`/groups/${groupId}/members/${userId}/mute`)
  return response.data.data
}

File: mobile/app/lib/core/models/group.dart
class GroupSummary {
  const GroupSummary({
    required this.groupId,
    required this.groupName,
    this.description,
    this.avatarUrl,
    this.notice,
    this.ownerId,
    this.inviteCode,
    this.chatEnabled = true,
    this.noticeUnread = false,
    required this.memberCount,
    required this.unreadCount,
    this.mentionUnreadCount = 0,
  });

  final int groupId;
  final String groupName;
  final String? description;
  final String? avatarUrl;
  final String? notice;
  final int? ownerId;
  final String? inviteCode;
  final bool chatEnabled;
  final bool noticeUnread;
  final int memberCount;
  final int unreadCount;
  final int mentionUnreadCount;

  factory GroupSummary.fromJson(Map<String, dynamic> json) {
    return GroupSummary(
      groupId: (json['groupId'] as num?)?.toInt() ?? 0,
      groupName: json['groupName']?.toString() ?? '',
      description: json['description']?.toString(),
      avatarUrl: json['avatarUrl']?.toString(),
      notice: json['notice']?.toString(),
      ownerId: (json['ownerId'] as num?)?.toInt(),
      inviteCode: json['inviteCode']?.toString(),
      chatEnabled: _boolish(json['chatEnabled'], defaultValue: true),
      noticeUnread: _boolish(json['noticeUnread']),
      memberCount: (json['memberCount'] as num?)?.toInt() ?? 0,
      unreadCount: (json['unreadCount'] as num?)?.toInt() ?? 0,
      mentionUnreadCount: (json['mentionUnreadCount'] as num?)?.toInt() ?? 0,
    );
  }

  static bool _boolish(Object? value, {bool defaultValue = false}) {
    if (value == null) return defaultValue;
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) return value == '1' || value.toLowerCase() == 'true';
    return defaultValue;
  }
}

class GroupMessage {
  const GroupMessage({
    required this.messageId,
    required this.groupId,
    required this.senderId,
    required this.senderNickname,
    this.senderAvatarUrl,
    required this.messageType,
    required this.content,
    this.fileId,
    this.replyToMessageId,
    this.replyPreview,
    this.mentionUserIds = const <int>[],
    this.mentionAll = false,
    this.fileName,
    this.fileMimeType,
    this.fileSize,
    required this.sentAt,
    required this.recallStatus,
  });

  final int messageId;
  final int groupId;
  final int senderId;
  final String senderNickname;
  final String? senderAvatarUrl;
  final String messageType;
  final String content;
  final int? fileId;
  final int? replyToMessageId;
  final String? replyPreview;
  final List<int> mentionUserIds;
  final bool mentionAll;
  final String? fileName;
  final String? fileMimeType;
  final int? fileSize;
  final String sentAt;
  final int recallStatus;

  factory GroupMessage.fromJson(Map<String, dynamic> json) {
    return GroupMessage(
      messageId: (json['messageId'] as num?)?.toInt() ?? 0,
      groupId: (json['groupId'] as num?)?.toInt() ?? 0,
      senderId: (json['senderId'] as num?)?.toInt() ?? 0,
      senderNickname: json['senderNickname']?.toString() ?? '',
      senderAvatarUrl: json['senderAvatarUrl']?.toString(),
      messageType: json['messageType']?.toString() ?? 'text',
      content: json['content']?.toString() ?? '',
      fileId: (json['fileId'] as num?)?.toInt(),
      replyToMessageId: (json['replyToMessageId'] as num?)?.toInt(),
      replyPreview: json['replyPreview']?.toString(),
      mentionUserIds: (json['mentionUserIds'] as List<dynamic>? ?? const [])
          .whereType<num>()
          .map((item) => item.toInt())
          .toList(),
      mentionAll: _boolish(json['mentionAll']),
      fileName: json['fileName']?.toString(),
      fileMimeType: json['fileMimeType']?.toString(),
      fileSize: (json['fileSize'] as num?)?.toInt(),
      sentAt: json['sentAt']?.toString() ?? '',
      recallStatus: (json['recallStatus'] as num?)?.toInt() ?? 0,
    );
  }

  static bool _boolish(Object? value) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) return value == '1' || value.toLowerCase() == 'true';
    return false;
  }
}

class GroupJoinRequestItem {
  const GroupJoinRequestItem({
    required this.requestId,
    required this.groupId,
    required this.requesterId,
    required this.requesterNickname,
    this.requesterAvatarUrl,
    this.requesterEmail,
    required this.status,
    this.message,
    this.createdAt,
  });

  final int requestId;
  final int groupId;
  final int requesterId;
  final String requesterNickname;
  final String? requesterAvatarUrl;
  final String? requesterEmail;
  final int status;
  final String? message;
  final String? createdAt;

  factory GroupJoinRequestItem.fromJson(Map<String, dynamic> json) {
    return GroupJoinRequestItem(
      requestId: (json['requestId'] as num?)?.toInt() ?? 0,
      groupId: (json['groupId'] as num?)?.toInt() ?? 0,
      requesterId: (json['requesterId'] as num?)?.toInt() ?? 0,
      requesterNickname: json['requesterNickname']?.toString() ?? '',
      requesterAvatarUrl: json['requesterAvatarUrl']?.toString(),
      requesterEmail: json['requesterEmail']?.toString(),
      status: (json['status'] as num?)?.toInt() ?? 0,
      message: json['message']?.toString(),
      createdAt: json['createdAt']?.toString(),
    );
  }
}

class GroupMemberItem {
  const GroupMemberItem({
    required this.userId,
    required this.email,
    required this.nickname,
    this.avatarUrl,
    required this.role,
    this.groupNickname,
    this.muteUntil,
    this.noticeReadAt,
    this.joinedAt,
  });

  final int userId;
  final String email;
  final String nickname;
  final String? avatarUrl;
  final int role;
  final String? groupNickname;
  final String? muteUntil;
  final String? noticeReadAt;
  final String? joinedAt;

  factory GroupMemberItem.fromJson(Map<String, dynamic> json) {
    return GroupMemberItem(
      userId: (json['userId'] as num?)?.toInt() ?? 0,
      email: json['email']?.toString() ?? '',
      nickname: json['nickname']?.toString() ?? '',
      avatarUrl: json['avatarUrl']?.toString(),
      role: (json['role'] as num?)?.toInt() ?? 0,
      groupNickname: json['groupNickname']?.toString(),
      muteUntil: json['muteUntil']?.toString(),
      noticeReadAt: json['noticeReadAt']?.toString(),
      joinedAt: json['joinedAt']?.toString(),
    );
  }
}

class GroupNotificationItem {
  const GroupNotificationItem({
    required this.notificationId,
    required this.groupId,
    this.actorId,
    this.targetUserId,
    required this.noticeType,
    required this.content,
    required this.createdAt,
  });

  final int notificationId;
  final int groupId;
  final int? actorId;
  final int? targetUserId;
  final String noticeType;
  final String content;
  final String createdAt;

  factory GroupNotificationItem.fromJson(Map<String, dynamic> json) {
    return GroupNotificationItem(
      notificationId: (json['notificationId'] as num?)?.toInt() ?? 0,
      groupId: (json['groupId'] as num?)?.toInt() ?? 0,
      actorId: (json['actorId'] as num?)?.toInt(),
      targetUserId: (json['targetUserId'] as num?)?.toInt(),
      noticeType: json['noticeType']?.toString() ?? '',
      content: json['content']?.toString() ?? '',
      createdAt: json['createdAt']?.toString() ?? '',
    );
  }
}

class GroupNoticeReadStat {
  const GroupNoticeReadStat({
    required this.readCount,
    required this.memberCount,
  });

  final int readCount;
  final int memberCount;

  factory GroupNoticeReadStat.fromJson(Map<String, dynamic> json) {
    return GroupNoticeReadStat(
      readCount: (json['readCount'] as num?)?.toInt() ?? 0,
      memberCount: (json['memberCount'] as num?)?.toInt() ?? 0,
    );
  }
}
